content/blog/introducao-orientacao-objetos.md

```diff
@@ -0,0 +1,381 @@
+```csharp
+public Class Palio : Automovel {
+ this.DataDeFabricacao => DateTime.Parse("2017-06-01");
+ this.CodigoDeReferenciaDoFabricante => "Fiat Palio Fire 1.0 8V (Flex) 4p 2017";
+ this.Modelo => "Fire";
+
+ Acelerar(){ 
+ Velocidade = Velocidade + 6.52;
+ //Aceleração média por segundo tomando como base velocidade 0-100 Km/h em 15s
+ }
+ Frear(){ 
+ /* Como o cálculo para frenagem é composto
+    de variantes diversas vamos utilizar
+    a mesma base de cálculo, 6.52 Km/h
+    */
+    Velocidade = Velocidade - 6.52;
+    }
+ 
+ }
+
+ Colidiu(){ }
+ }
+```
+<small>Listagem 2: Classe Palio</small>
+
+```csharp
+Class Program {
+ public static void Main( ) {
+ Automovel veiculo = new Palio(); // Instancia de um veículo implementando a API
+
+ int segundosParaTestar = 5, // Por quanto tempo o programa ira rodar
+     segundosPassadosDoTeste = 0; // Armazena o tempo percorrido pelo teste
+
+ while (segundosParaTestar > segundosPassadosDoTeste) // Enquanto o tempo total não é atingido
+ {
+  veiculo.Acelerar(); // Acelera o veiculo
+  Thread.Sleep(1000); // Pelo tempo de 1 segundo
+  segundosPassadosDoTeste = segundosPassadosDoTeste + 1; // Incrementa em 1 segundo o tempo do teste
+ }
+ 
+ float velocidadeNoInstante = veiculo.Velocidade; // Armazena a velocidade do veiculo após o tempo desejado
+ 
+ while (veiculo.Velocidade > 0)
+  veiculo.Frear(); //Para o automóvel
+ 
+ }
+```
+<small>Listagem 3: Programa para teste de aceleração</small>
+
+O encapsulamento foi descrito como: a espera do comportamento especificado para o consumidor da API. Nesse caso o programa utiliza uma instância de um Veiculo, nesse caso uma classe Palio que implementa a API. 
+
+De acordo com a especificação do próprio pálio, a classe Palio abstraí as informações do veículo em algo que faça sentido ao domínio da aplicação e encapsula o comportamento de aceleração e frenagem, assim o consumidor pode utilizar a API e não se preocupar em implementar o comportamento de aceleração e frenagem do pálio.
+
+Para mudar o comportamento do programa para outro veículo é muito fácil, basta criar uma nova herança e através do polimorfismo fazer as chamadas a API disponível na classe Automovel e nosso programa pode se beneficiar do encapsulamento proporcionado e já se adequar a uma ferrari.
+
+### Herança e Polimorfismo
+
+O conceito de herança é algo muito forte na literatura do paradigma orientado a objetos, surgiu como forma de centralização e reuso de propriedades e comportamentos.
+
+Herança em OOP é a capacidade de uma estrutura ter os mesmos comportamentos de suas estruturas bases. Ou analogamente, é a capacidade de herdar algo dos ancestrais.
+
+Mas porquê ancestrais?
+
+Pois precisamos ter o contrato (contrato é o mesmo que API) da estrutura antes de fazer uma estrutura descendente, ou herança.
+
+E Polimorfismo?
+
+É a capacidade de uma estrutura descendente assumir a forma de um ancestral em um contexto, sem que isso interfira no resultado.
+
```

Comment:
Não creio que isso está correto.
Definição em inglêses:
Is the characteristic of being able to assign a different meaning or usage to something in different contexts - specifically, to allow an entity such as a variable, a function, or an object to have more than one form.

Há também várias formas de polimorfismo e talvez fosse interessante escrever sobre elas.